crates/wasmtime/src/runtime/component/concurrent.rs

```diff
@@ -2816,54 +2906,240 @@ impl Instance {
+            if state.get_mut(thread.task)?.ready_to_delete() {
+                Waitable::Guest(thread.task).delete_from(state)?;
+            }
+            log::trace!("thread start: restored {old_thread:?} as current thread");
+
+            Ok(())
+        });
+        let guest_call = WorkItem::GuestCall(GuestCall {
+            thread,
+            kind: GuestCallKind::StartExplicit(start_func),
+        });
+        if high_priority {
+            store
+                .0
+                .concurrent_state_mut()
+                .push_high_priority(guest_call);
+        } else {
+            store.0.concurrent_state_mut().push_low_priority(guest_call);
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn resume_suspended_thread<T: 'static>(
+        self,
+        mut store: StoreContextMut<T>,
+        runtime_instance: RuntimeComponentInstanceIndex,
+        thread_id: TableId<GuestThread>,
+        high_priority: bool,
+    ) -> Result<()> {
+        let state = store.0.concurrent_state_mut();
+        let guest_thread = QualifiedThreadId::qualify(state, thread_id)?;
+        let thread = state.get_mut(guest_thread.thread)?;
+
+        match mem::replace(&mut thread.state, GuestThreadState::Running) {
+            GuestThreadState::NotStartedExplicit {
+                start_func_table_idx,
+                start_func_idx,
+                context,
+            } => {
+                self.start_thread(
+                    &mut store,
+                    guest_thread,
+                    runtime_instance,
+                    start_func_table_idx,
+                    start_func_idx,
+                    context,
+                    high_priority,
+                )?;
+            }
+            GuestThreadState::Suspended(fiber) => {
+                log::trace!("resuming thread {thread_id:?} that was suspended");
+                if high_priority {
+                    store
+                        .0
+                        .concurrent_state_mut()
+                        .push_high_priority(WorkItem::ResumeFiber(fiber));
                 } else {
-                    false
+                    store
+                        .0
+                        .concurrent_state_mut()
+                        .push_low_priority(WorkItem::ResumeFiber(fiber));
                 }
-            })
+            }
+            _ => {
+                bail!("cannot resume thread which is not suspended");
+            }
+        }
+        Ok(())
+    }
+
+    /// Helper function for the `thread.yield`, `thread.yield-to`, `thread.suspend`,
+    /// and `thread.switch-to` intrinsics.
+    pub(crate) fn suspension_intrinsic<T: 'static>(
+        self,
+        mut store: StoreContextMut<T>,
+        caller: RuntimeComponentInstanceIndex,
+        cancellable: bool,
+        yielding: bool,
+        to_thread: Option<u32>,
+    ) -> Result<bool> {
```

Comment:
Change to a `Result<Something>` where `Something` represents `Cancelled` or `NotCancelled`